Dnf5/feature/config

jrohel opened the pull request.

jrohel committed:
[dnf5] Add base exception support

jrohel committed:
[dnf5] Add fmt library for text formating

Formatting library for C++.
Safe printf implementation including the POSIX extension for positional
arguments and implementation of C++20 std::format.

Idea:
If we only use C++20 style (std::format) in libdnf then the fmt library
will only be needed for compilers without C++20 support.

jrohel committed:
[dnf5] Modify Option class for new libdnf

Uses a new exception hierarchy, snake_case, ...

jrohel committed:
[dnf5] Modify implementation of basic config option types for new libdnf

jrohel committed:
[dnf5] Modify additional config option types for the new libdnf

OptionEnum, OptionStringList

jrohel committed:
[dnf5] Modify OptionChild class for new libdnf

jrohel committed:
[dnf5] Modify OptionSeconds for the new libdnf

jrohel committed:
[dnf5] Modify OptionPath for new libdnf

jrohel committed:
[dnf5] Modify OptionBinds for new libdnf

jrohel committed:
[dnf5] Modify Config for new libdnf

jrohel committed:
[dnf5] Modify ConfigMain for new libdnf

jrohel committed:
[dnf5] Add LogRouter::swap_logger() method

void swap_logger(std::unique_ptr<Logger> & logger, size_t index);
Swaps the logger at the "index" position with another "logger".

It is used, for example, to replace a temporary (memory buffer) logger
with a final one.

jrohel committed:
[dnf5] Add LogRouter into Base class

jrohel committed:
[dnf5] microdnf - add Base and set Logger

Conan-Kudo commented on include/libdnf/conf/config_main.hpp:
```diff
@@ -0,0 +1,187 @@
+    OptionStringList & protected_packages();
+    OptionString & username();
+    OptionString & password();
+    OptionBool & gpgcheck();
```
Can we go ahead and lay the groundwork here for the GPG check refactor? (As noted in [RhBug:1433592#c7](https://bugzilla.redhat.com/show_bug.cgi?id=1433592#c7)). This is the perfect time to do so.

j-mracek replied:
@Conan-Kudo Am I correct that this is a request to drop RHEL7, RHEL8 compatibility in favour of Zypper behaviour?

Conan-Kudo replied:
Somewhat. The `gpgcheck_policy` option would control whether we behave as YUM does or as Zypper does for `gpgcheck`. In general, I think the correct default is to behave like Zypper here, but I'm okay with making the default the same as YUM, as long as we can tune this.

Conan-Kudo commented on include/libdnf/conf/config_main.hpp:
```diff
@@ -0,0 +1,187 @@
+    OptionPath & destdir();
+    OptionString & comment();
+    OptionBool & downloadonly();
+    OptionBool & ignorearch();
```
Let's go ahead and add the sticky vendor options that are part of #907.

j-mracek replied:
We have to first finish review of #907.
We would like to develop dnf5 into state of libdnf master, then we will backport all additional patches from dnf-4

Conan-Kudo commented on include/libdnf/conf/const.hpp:
```diff
@@ -0,0 +1,48 @@
+constexpr const char * PERSISTDIR = "/var/lib/dnf";
+constexpr const char * SYSTEM_CACHEDIR = "/var/cache/dnf";
+
+constexpr const char * CONF_FILENAME = "/etc/dnf/dnf.conf";
```
Where is the `/etc/dnf/conf.d` setup?

jrohel replied:
Thanks for comments. I wrote notes. The code is under development. I'm trying to reuse old code and adopt it to libdnf5.

j-mracek replied:
Why '/etc/dnf/conf.d' ?

Conan-Kudo replied:
That's what's in the `dnf` PR: https://github.com/rpm-software-management/dnf/pull/1414

Conan-Kudo reviewed (changes requested):
Generally looks like a good start, but there are a few things I feel should be addressed...

jrohel committed:
[dnf5] Add rpm::ConfigRepo (stores configuration of RPM repository)

jrohel committed:
[dnf5] Add ConfigMain into Base class

jrohel committed:
[dnf5] Base: Add methods for main configuration loading

jrohel committed:
[dnf5] microdnf - load main configuration

jrohel committed:
[dnf5] Move bgettext to the new location

Support for text translation

jrohel committed:
[dnf5] Move tinyformat to the new location

Type safe printf() replacement

jrohel committed:
[dnf5] Fix: tinyformat - FormatListN template type

error: conversion from ‘long unsigned int’ to ‘int’ may change value [-Werror=conversion]
 1017 | detail::FormatListN<sizeof...(Args)> makeFormatList(const Args&... args)

jrohel committed:
[dnf5] cmake - set gettext domain for translations

jrohel committed:
[dnf5] Add GLib2, GPGme, and librepo libraries

We don't want to depend on GLib, but librepo and libmodulemd use it
in  API.

jrohel committed:
[dnf5] Move iniparser from old libdnf - parser for dnf ini files

jrohel committed:
[dnf5] Modify iniparser for new libdnf

jrohel committed:
[dnf5] Move PreserveOrderMap from old libdnf

jrohel committed:
[dnf5] Modify PreserveOrderMap for new libdnf

jrohel committed:
[dnf5] Move ConfigParser from old libdnf

jrohel committed:
[dnf5] Modify ConfigParser for new libdnf

jrohel committed:
[dnf5] Add utils - have_files_same_content function

j-mracek commented on include/libdnf/utils/exception.hpp:
```diff
@@ -0,0 +1,75 @@
+    virtual const char * get_domain_name() const noexcept { return "libdnf"; }
+    virtual const char * get_name() const noexcept { return "Exception"; }
+    virtual const char * get_description() const noexcept { return ""; }
+};
```
Shell we extend all errors by error code value?

jrohel replied:
Example:
An `Exception` with error code `bad_value` can be replaced by inherited specialized exception `BadValue`.

I thought about adding error code for easy transform to other languages eg C. The same approach as get_name() -> Each exception class will have unique `id` (`error code`). But how create well defined unique `id`? The string `get_domain_name() + get_name()` is unique to each exception now.

Note:
There is `error_code` in my SystemError exception class. Because system error codes are defined and I think that is no need to create specialized exception for each system error code.
I have some concept but it can be modified.

jrohel committed:
[dnf5] Move Repo from old libdnf

jrohel committed:
[dnf5] Modify rpm repository support (rpm::Repo) for new libdnf

jrohel committed:
[dnf5] WeakPtr implementation

WeakPtr is a "smart" pointer. It contains a pointer to resource and to
guard of resource. WeakPtr pointer can be owner of the resource. However,
the resource itself may depend on another resource.
WeakPtr registers/unregisters itself at the guard of resource. And
the resource guard invalidates the registered WeakPtrs when the resource
is unusable (eg. its dependecny was released).

jrohel committed:
[dnf5] Add Set class

jrohel committed:
[dnf5] Add Query support

jrohel committed:
[dnf5] Add common Sack

jrohel committed:
[dnf5] Implement rpm RepoQuery

jrohel committed:
[dnf5] Implement rpm RepoSack

jrohel committed:
[dnf5] microdnf: Add substitutions and repos loading, gpg import, ...

Only example will be updated.

jrohel committed:
[dnf5] Set class: Add operators | & - ^

jrohel committed:
[dnf5] Support creating rpm repos from config file and directories

jrohel committed:
[dnf5] Add rpm_repo_sack into Base

jrohel committed:
[dnf5] Load conf files from dir in alphabetical order

jrohel committed:
[dnf5] Base: Add substitutions map

jrohel committed:
[dnf5] Move Option to the new location

Abstract class. Base class for all configuration options type.

jrohel committed:
[dnf5] Move OptionChild template class to the new location

The type of option is taken from linked option during compile type.
It uses topical default value and parameters from linked option during
runtime.

jrohel committed:
[dnf5] Move basic config option types to the new location

OptionBool, OptionNumber, OptionString

jrohel committed:
[dnf5] Move additional config option types to the new location

OptionEnum, OptionStringList

jrohel committed:
[dnf5] Move OptionSeconds to the new location

option that represents a time duration

jrohel committed:
[dnf5] Move OptionPath to the new location

option for storing file/directory path

jrohel committed:
[dnf5] Move OptionBinds to the new location

It maps option names (text names readed from config file, command line, ...)
to option objects.

jrohel committed:
[dnf5] Move Config to the new location

Base class for configuration

jrohel committed:
[dnf5] Move ConfigMain to the new location

Stores main configuration of libdnf
